### PR TITLE
Speed up `parseDateTime[InJodaSyntax]Or(Null|Zero)` on invalid inputs

### PR DESCRIPTION
**Pull request by liuneng1994**
Thread synchronization is implicit in the process of Exception generation. When a large amount of data is illegal, a lot of exceptions will be generated concurrently, which makes the performance very bad. By optimizing error handling, the concurrency performance in this situation can be improved.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions `parseDateTimeOrNull`, `parseDateTimeOrZero`, `parseDateTimeInJodaSyntaxOrNull` and `parseDateTimeInJodaSyntaxOrZero` are now up to 10x(1000x in ci) faster on invalid inputs.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by rschu1ze** [comment]
Runtime of the the included performance tests improves from 11.3 sec to 1.1 sec (good!).

The PR itself is unfortunately too messy:
- Please format `parse_illegal_datetime.xml` properly. Kindly see tests/performance/dotProduct.xml for an example.
- As for macro names, please use `INSTRUCTION` instead of `INSTURCION`.
- Stateless tests for `parseDateTime` failed locally for me. I'll start CI, so we will see what fails in detail.
- There is zero code documentation. The code is becoming more complex (in particular custom error handling is added) and future readers of the code will have no clue why all of that extra complexity was added. Please add ample code documentation.
- It would be much nicer if we could C++ (lambda) functions instead of macros.

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 1e691e71e07b0cb7cce00baa1aae60fcd329468b with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62538/1e691e71e07b0cb7cce00baa1aae60fcd329468b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by liuneng1994** [comment]
> Runtime of the the included performance tests improves from 11.3 sec to 1.1 sec (good!).
> 
> The PR itself is unfortunately too messy:
> 
> * Please format `parse_illegal_datetime.xml` properly. Kindly see tests/performance/dotProduct.xml for an example.
> * As for macro names, please use `INSTRUCTION` instead of `INSTURCION`.
> * Stateless tests for `parseDateTime` failed locally for me. I'll start CI, so we will see what fails in detail.
> * There is zero code documentation. The code is becoming more complex (in particular custom error handling is added) and future readers of the code will have no clue why all of that extra complexity was added. Please add ample code documentation.
> * It would be much nicer if we could C++ (lambda) functions instead of macros.

ok, I will modify it, but lambda expressions may not achieve the effect of code reuse, because here we need to judge the status of the return value and decide whether to exit the current function. It seems that only macros can do it.

**Comment by liuneng1994** [comment]
image
The fundamental reason is that Exception will obtain the stack of the current thread. Another way to modify it is to introduce a lighter Exception type, but it is still possible to mix it and it is not easy to detect. After all, a benchmark will not be written for all error conditions.

**Comment by liuneng1994** [comment]
image
Unbelievable, a thousand times

File: src/Functions/parseDateTime.cpp [hunk 1]
```diff
@@ -105,40 +105,65 @@ namespace
...
         std::variant<Int32, Int64, Pos> res;
 
+        /// Helper to disambiguate the first Status argument: Pos, Int64, Int32 (for success) vs. int (for errors)
+        struct IsValueTag {};
+
         Status() = default;
-        Status(Pos pos_) : res(pos_) { }
-        Status(Int64 pos_) : res(pos_) { }
-        Status(Int32 pos_) : res(pos_) { }
+        Status(IsValueTag /*tag*/, Pos pos_) : res(pos_) { }
```
Comment (liuneng1994) [inline src/Functions/parseDateTime.cpp:112]: Good method. I learned it. I have been struggling with how to distinguish error code and result value.

File: src/Functions/parseDateTime.cpp [hunk 2]
```diff
@@ -785,7 +776,8 @@ namespace
             static Status readNumber2(Pos cur, Pos end, [[maybe_unused]] const String & fragment, T & res)
             {
                 if constexpr (need_check_space == NeedCheckSpace::Yes)
-                    SAFE_EXECUTE_VOID_FUNCTION(checkSpace(cur, end, 2, "readNumber2 requires size >= 2", fragment))
+                    if (Status status = checkSpace(cur, end, 2, "readNumber2 requires size >= 2", fragment); status.isError())
```
Comment (liuneng1994) [inline src/Functions/parseDateTime.cpp:799]: Also learned a new way

**Comment by rschu1ze** [comment]
@liuneng1994 I like the direction of this PR (thanks again for providing it). What I don't like is the added boilerplate code. The `Status` class is basically a weird re-implementation of `std::expected` ([docs](https://en.cppreference.com/w/cpp/utility/expected)) - both your initial draft as well as my improved version. We should not re-implement stuff which the standard library provides.

I made a draft how a re-implementation of this PR with std::expected could look like: https://github.com/ClickHouse/ClickHouse/pull/62607

It has one technical caveat: ClickHouse is compiled with C++23, but it's standard library (aka. submodule contrib/llvm-project/libcxx) is currently stuck at the LLVM 15 version, whereas only standard libraries shipped with LLVM 16 come with the required header `<expected>`. My PR therefore imports https://github.com/TartanLlama/expected as a new submodule which is a standalone and C++-standard compatible implementation of `<expected>`. This is an acceptable temporary solution until we upgraded to LLVM 16.

So my humble ask would be if you like to check my draft, re-implement the stuff in this PR on top of it (please make a new PR for that), push, and please set me as a reviewer?

This is some extra effort (sorry) but in the long run, the code quality will benefit if we don't re-invent the wheel.

**Comment by liuneng1994** [comment]
> @liuneng1994 I like the direction of this PR (thanks again for providing it). What I don't like is the added boilerplate code. The `Status` class is basically a weird re-implementation of `std::expected` ([docs](https://en.cppreference.com/w/cpp/utility/expected)) - both your initial draft as well as my improved version. We should not re-implement stuff which the standard library provides.
> 
> I made a draft how a re-implementation of this PR with `std::expected` could look like: #62607
> 
> It has one technical caveat: ClickHouse is compiled with C++23, but it's standard library (aka. submodule contrib/llvm-project/libcxx) is currently stuck at the LLVM 15 version, whereas only standard libraries shipped with LLVM 16 come with the required header `<expected>`. My PR therefore imports https://github.com/TartanLlama/expected as a new submodule which is a standalone and C++-standard compatible implementation of `<expected>`. This is an acceptable temporary workaround until we upgraded to LLVM 16.
> 
> So my humble ask would be if you like to check my draft, re-implement the stuff in this PR on top of it (please make a new PR for that), push, and please set me as a reviewer?
> 
> This is some extra effort (sorry) but in the long run, the code quality will benefit if we don't re-invent the wheel.

No problem, I will re-implement a new version based on #62607 and submit a new PR. Thank you for your review, I have learned a lot of new knowledge.

**Pull request closed without merging**